extensions/data-transfer/portability-data-transfer-flickr/src/main/java/org/dataportabilityproject/datatransfer/flickr/photos/FlickrPhotosImporter.java

```diff
@@ -80,46 +83,64 @@ public ImportResult importItem(UUID jobId, AuthData authData, PhotosContainerRes
+      importAlbums(data.getAlbums(), tempPhotosData);
+      jobStore.update(jobId, tempPhotosData);
     }
-    jobStore.update(jobId, tempPhotosData);
-
-    for (PhotoModel photo : data.getPhotos()) {
-      try {
-        String photoId = uploadPhoto(photo);
-        String oldAlbumId = photo.getAlbumId();
-        TempPhotosData tempData = jobStore.findData(TempPhotosData.class, jobId);
-        String newAlbumId = tempData.lookupNewAlbumId(oldAlbumId);
-        if (Strings.isNullOrEmpty(newAlbumId)) {
-          // This means that we havent created the new album yet, create the photoset
-          PhotoAlbum album = tempData.lookupAlbum(CACHE_ALBUM_METADATA_PREFIX + oldAlbumId);
-          Photoset photoset =
-              photosetsInterface.create(
-                  COPY_PREFIX + album.getName(), album.getDescription(), photoId);
-          tempData.addAlbumId(oldAlbumId, photoset.getId());
-        } else {
-          // We've already created a new album, add the photo to the new album
-          photosetsInterface.addPhoto(newAlbumId, photoId);
+
+    if (data.getPhotos() != null) {
+      for(PhotoModel photo : data.getPhotos()) {
+        try {
+          importSinglePhoto(jobId, photo);
+        } catch (FlickrException | IOException e) {
+          return new ImportResult(ResultType.ERROR, "Error importing photo " + e.getMessage());
         }
-        jobStore.update(jobId, tempData);
-      } catch (FlickrException | IOException e) {
-        // TODO: figure out retries
-        return new ImportResult(
-            ImportResult.ResultType.ERROR, "Error importing item: " + e.getMessage());
       }
     }
+
     return new ImportResult(ImportResult.ResultType.OK);
   }
 
+  // Store any album data in the cache because Flickr only allows you to create an album with a
+  // photo in it, so we have to wait for the first photo to create the album
+  private void importAlbums(Collection<PhotoAlbum> albums, TempPhotosData tempPhotosData) {
+    for (PhotoAlbum album : albums) {
+      tempPhotosData.addAlbum(CACHE_ALBUM_METADATA_PREFIX + album.getId(), album);
+    }
+  }
+
+  private void importSinglePhoto(UUID id, PhotoModel photo)
+      throws FlickrException, IOException {
+      String photoId = uploadPhoto(photo);
+
+      // TODO: what happens when the photo isn't associated with an album?
+      String oldAlbumId = photo.getAlbumId();
+      TempPhotosData tempData = jobStore.findData(TempPhotosData.class, id);
+      String newAlbumId = tempData.lookupNewAlbumId(oldAlbumId);
+      if (Strings.isNullOrEmpty(newAlbumId)) {
+        // This means that we havent created the new album yet, create the photoset
+        PhotoAlbum album = tempData.lookupAlbum(CACHE_ALBUM_METADATA_PREFIX + oldAlbumId);
+        Photoset photoset =
+            photosetsInterface.create(
+                COPY_PREFIX + album.getName(), album.getDescription(), photoId);
+        tempData.addAlbumId(oldAlbumId, photoset.getId());
```

Comment:
Once we add the actual album mapping, we should remove the version with the CACHE_ALBUM_METADATA_PREFIX

Reply:
Done